### DISCUSSION TITLE
How to run new shell scripts in current execution context Windows

### DISCUSSION THREAD

**Discussion by ta32** [nushell/nushell#9225] [Q&A]
in PowerShell when I execute my scripts they run in the same execution context. 

to have this behavior you need to invoke your scripts like this:
`source myscript.nu `

If you are in PowerShell and you call "somescript" - if "somescript.ps1" is on your path it will execute the script in the current context and will set environment variables 

Is there a way to have a similar behavior in Nu?

Making .nu scripts executable allows you to invoke them but they run in a different context
```
setx PATHEXT "%PATHEXT%;.nu;" /m
assoc .nu=NushellScript
ftype NushellScript="C:\Program Files\nu\bin\nu.exe" %1 %*
```

**Comment by fdncred** [comment]
> When I invoke my shell script they still don't set environment variables in the execution context.

If you're executing scripts with `nu -c` it's shelling out of your current shell, launching nushell, running the commands mentioned after `-c`, and then exiting. So, your env vars are being set probably just fine, but you're exiting so you can see them.

I thought in powershell you had to do something like `& somescript.ps1` to get it to run in the current context. That's kind of similar to `source some_script.nu` or `use some_module.nu`.

If nushell is your shell all the time, then you could just source and not do nu -c.

If you're looking for a way to do nu -c and yet still keep env vars available in whatever the parent shell is, i'm not sure that's possible. the parent shell would have to know about nushell's environment and somehow nushell would have to pass those env vars to the parent shell. i'm not quite sure how to do that or if it's possible since every process has it's own copy of the environment, at least on windows it does.

**Reply by ta32** [reply]
Sorry my post was wrong. PATHEXT is only required if you want to execute your PowerShell scripts from explorer or from a different shell - in which they will execute in a separate context. ( I forgot about this )

I setup a vm without these settings and ran the test

**The Scripts directory is in the path** 
![image](https://github.com/nushell/nushell/assets/26617779/6352d975-8f21-4cb6-ae45-b1a761be928c)

testn.nu
```
print "Hello"
$env.E = 'e'
print "World"
```

testp.ps1
```
echo "hello"
$env:D = "world"
echo $PSVersionTable
```

I can execute a script - calling it by name ( if its on the path )
![image](https://github.com/nushell/nushell/assets/26617779/9c40a2e4-634f-4950-924f-44e612f2d49b)

In nushell when trying to invoke the script
![image](https://github.com/nushell/nushell/assets/26617779/ca3a98b5-a788-44d7-9655-ce8969e0a9ec)

This is why I thought I needed to make it executable (PATHEXT) - but you don't have to do this in PowerShell.

In nushell you need to source your script with the absolute path like 
`source C:\Users\Administrator\Downloads\Scripts\testn.nu`

but doing this is inconvenient.

In PowerShell if the script is on your PATH you can call its name from anywhere and it will execute in the current context. 

making .nu files executables and putting them on the PATH - half results in this behavior. I assume this is by design.

**Comment by ta32** [comment]
https://nushell.sh/book/how_nushell_code_gets_run.html#_1-sourcing-a-dynamic-path
Is this the reason why we can run/source scripts that are on the PATH?
I still fully don't understand the restriction.

The work around I did on windows was to make .nu extension executable. So when a user in _any_ shell try's to call testn.nu ( assuming its on the path) windows will try to execute it by calling nu.exe ...

I tested the behavior on WSL and when you include Shebangs- its equivalent to the behavior you get when you  make .nu executable on windows using PATHEXT.

On windows - your .nu scripts can be on your PATH and calling it by name will execute the script ( in a child process ) from any directory. But this behavior didn't work in WSL2. But I suspect this is how windows works when searching for external programs.
